soup-ranking.go

```diff
@@ -4,41 +4,70 @@ import (
 	"log"
 	"net/http"
 	"os"
+	"strings"
 
-	"github.com/gorilla/handlers"
 	"github.com/gorilla/mux"
 	"github.com/sjuls/soup-ranking/dbctx"
-	"github.com/sjuls/soup-ranking/routes"
+	"github.com/sjuls/soup-ranking/middleware"
+	"github.com/sjuls/soup-ranking/score"
+	"github.com/sjuls/soup-ranking/slack"
+	"github.com/sjuls/soup-ranking/soup"
+	"github.com/sjuls/soup-ranking/status"
 )
 
 var (
-	allowedOrigins = []string{
-		"https://junesoup.surge.sh",
-		"http://junesoup.surge.sh",
-		"https://soup-ranking.herokuapp.com",
-		"http://soup-ranking.herokuapp.com",
+	middlewares = [](func(handler http.Handler) http.Handler){
+		middleware.Log,
+		middleware.CORS,
 	}
 )
 
 func main() {
 	port := os.Getenv("PORT")
 	database := os.Getenv("DATABASE_URL")
+	slackVerificationToken := os.Getenv("SLACK_VERIFICATION_TOKEN")
+	slackAccessToken := os.Getenv("SLACK_ACCESS_TOKEN")
+	slackBaseURL := os.Getenv("SLACK_BASEURL")
+	slackAdminUsers := os.Getenv("SLACK_ADMIN_USERS")
 
-	if err := dbctx.Init(&database); err != nil {
+	connFactory, err := dbctx.NewConnectionFactory(&database)
+	if err != nil {
 		panic(err)
 	}
 
+	scoreRepository := score.NewRepository(connFactory)
+	soupRepository := soup.NewRepository(connFactory)
+
 	router := mux.NewRouter().StrictSlash(true)
-	routes.AddStatus(router)
-	routes.AddScore(router)
+	routes := []func(router *mux.Router){
+		status.AddRoute,
+		score.AddRoute(
+			scoreRepository,
+		),
+		slack.AddRoute(
+			slackVerificationToken,
+			slackBaseURL,
+			slackAccessToken,
+			scoreRepository,
+			strings.Split(slackAdminUsers, ","),
+		),
+	}
+
+	registerRoutes(router, routes)
 
-	log.Fatal(http.ListenAndServe(":"+port, wrapCORS(router)))
+	log.Printf("Starting up soup-ranking on port %s", port)
+	log.Fatal(http.ListenAndServe(":"+port, applyMiddleware(router)))
 }
 
-func wrapCORS(router *mux.Router) http.Handler {
-	corsHeaders := handlers.AllowedHeaders([]string{"X-Requested-With", "Content-Type"})
-	corsOrigin := handlers.AllowedOrigins(allowedOrigins)
-	corsMethods := handlers.AllowedMethods([]string{"GET", "HEAD", "POST", "PUT", "OPTIONS"})
+func registerRoutes(router *mux.Router, routes []func(router *mux.Router)) {
+	for _, route := range routes {
+		route(router)
+	}
+}
 
-	return handlers.CORS(corsHeaders, corsOrigin, corsMethods)(router)
+func applyMiddleware(handler http.Handler) http.Handler {
```

Comment:
I guess one could make this return an interface which is an `http.Handler` as well as an `thener` (you find a good name), so that middlewares could be chained like `applyMiddleware(router).then(login).then(wowSuchMiddleware).handler`

Reply:
That at least would make it easier to diverge from the list of `middlewares`